File: 프로그래머스/3/43164. 여행경로/여행경로.cs
using System;
using System.Collections.Generic;

public class Solution {
    private List<string> result;
    private bool[] visited;
    public string[] solution(string[,] tickets) {
        result = new List<string>();
        visited = new bool[tickets.GetLength(0)];
    
        BackTracking(tickets, 0, "ICN", "ICN");
        
        result.Sort();
    
        
        string[] answer = result[0].Split(' ');
        return answer;
    }
    
    private void BackTracking(string[,] tickets, int depth, string start, string route)
    {
        if(depth == tickets.GetLength(0))
        {
            result.Add(route);
            return;
        }
        
        for(int i = 0; i < tickets.GetLength(0); i++)
        {
            if(tickets[i, 0].Equals(start) && !visited[i])
            {
                visited[i] = true;
                BackTracking(tickets, depth + 1, tickets[i, 1], route + " " + tickets[i, 1]);
                visited[i] = false;
            }
        }
    }
}

File: 프로그래머스/3/43164. 여행경로/여행경로.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

vector<bool> visited;
vector<string> answer;
bool fin = false;

void dfs(string cur, vector<vector<string>> tickets, vector<string> path, int cnt)
{
    if(fin) return;
    path.push_back(cur);
    
    if(cnt == tickets.size())
    {
        answer = path;
        fin = true;
        return;
    }
    
    for(int i = 0; i < tickets.size(); i++)
    {
        if(!visited[i] && tickets[i][0] == cur)
        {
            visited[i] = true;
            dfs(tickets[i][1], tickets, path, cnt + 1);
            visited[i] = false;
        }
    }
    path.pop_back();
}

vector<string> solution(vector<vector<string>> tickets) {
    
    sort(tickets.begin(), tickets.end());
    visited.resize(tickets.size(), false);
    
    vector<string> path;
    dfs("ICN", tickets, path, 0);
    
    return answer;
}